### PR TITLE
Fix showing SQL LIMIT exceptions to the user  [HZ-1436]

### LINKED ISSUE #22024: SEVERE Exception for SQL limit clause [HZ-1436]
**Describe the bug**
An exception appears in the logs for an SQL clause that stops because of LIMIT

**Expected behavior**
No exception in logs, noise

**To Reproduce**
```
	public static void main(String[] args) throws Exception {
		Config config = new Config();
		config.getJetConfig().setEnabled(true);
		
		HazelcastInstance hazelcastInstance = Hazelcast.newHazelcastInstance(config);
		
		hazelcastInstance.getMap("neil").put("Hello", "World");
		hazelcastInstance.getMap("neil").put("Goodbye", "World");
		
		String mapping = "CREATE MAPPING neil TYPE IMap " + " OPTIONS ( " + " 'keyFormat' = 'java',"
                + " 'keyJavaClass' = '" + String.class.getName() + "'," + " 'valueFormat' = 'java',"
                + " 'valueJavaClass' = '" + String.class.getName() + "'" + " )";
		
		String query = "SELECT * FROM neil LIMIT 1";
		
		List.of(mapping, query).forEach(sql -> hazelcastInstance.getSql().execute(sql));
	}
```

gives
```
com.hazelcast.jet.JetException: Execution on a member failed: com.hazelcast.sql.impl.ResultLimitReachedException: Done by reaching the item number in SQL LIMIT clause
        at com.hazelcast.jet.impl.LightMasterContext.findError(LightMasterContext.java:310)
        at com.hazelcast.jet.impl.LightMasterContext.lambda$createContext$2(LightMasterContext.java:164)
        at com.hazelcast.jet.impl.LightMasterContext.lambda$invokeOnParticipant$8(LightMasterContext.java:281)
        at com.hazelcast.spi.impl.AbstractInvocationFuture$WhenCompleteNode.lambda$execute$0(AbstractInvocationFuture.java:1564)
        at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Caused by: com.hazelcast.sql.impl.ResultLimitReachedException: Done by reaching the item number in SQL LIMIT clause

```

**Additional context**
Hazelcast Platform 5.2-SNAPSHOT (20220819 - 111323f)

### LINKED ISSUE #22254: Suspending a job prints the following error log with an exception into the log [HZ-1484]
**Describe the bug**

Suspending a job prints the following error log with an exception into the log.

**Expected behavior**

No exception is printed. (This is the state in 5.1.4).

**To Reproduce**

1. Submit any streaming job (e.g. hello-world job from https://github.com/hazelcast/hazelcast-code-samples/tree/master/jet/hello-world)

2. Suspend the job
```
bin/hz-cli suspend hello-world
```

3. In the log there is the following exception:
```
2022-09-16 11:20:16,267 [ERROR] [ForkJoinPool.commonPool-worker-9] [c.h.j.i.o.StartExecutionOperation]: [172.17.0.2]:5701 [dev] [5.2-SNAPSHOT] null
com.hazelcast.jet.impl.exception.TerminatedWithSnapshotException: null
	at com.hazelcast.jet.impl.execution.ExecutionContext.lambda$beginExecution$3(ExecutionContext.java:240) ~[hazelcast-5.2-SNAPSHOT.jar:5.2-SNAPSHOT]
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073) ~[?:?]
	at com.hazelcast.jet.impl.util.NonCompletableFuture.internalComplete(NonCompletableFuture.java:68) ~[hazelcast-5.2-SNAPSHOT.jar:5.2-SNAPSHOT]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$ExecutionTracker.taskletDone(TaskletExecutionService.java:496) ~[hazelcast-5.2-SNAPSHOT.jar:5.2-SNAPSHOT]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.dismissTasklet(TaskletExecutionService.java:429) ~[hazelcast-5.2-SNAPSHOT.jar:5.2-SNAPSHOT]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.runTasklet(TaskletExecutionService.java:406) ~[hazelcast-5.2-SNAPSHOT.jar:5.2-SNAPSHOT]
	at java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:807) ~[?:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.run(TaskletExecutionService.java:369) ~[hazelcast-5.2-SNAPSHOT.jar:5.2-SNAPSHOT]
	at java.lang.Thread.run(Thread.java:829) ~[?:?]
```

**Additional context**

Latest 5.2-SNAPSHOT (85fcced)

Marking as critical because of the amount of support tickets / questions that would be raised if we leave this unattended.

### PR DESCRIPTION
**Pull request by TomaszGaweda**
Partially eliminates, partially hides exception-driven flow in SQL queries' limit.

Basic algorithm:
- if a job is a batch job, then emitting last item in DoneTracker will cause inbox to be cleared and in next step machine step it will run `complete` method, which will return true and stop processing. Iterator will return `hasNext=false` without any exception.
- if a job is a streaming job, an exception will be thrown. However, new utility isTechnicalException will guard that such exceptions won't appear in logs.

If there was an exception during SQL processing, it will be rethrown and semantics will be the same as before - job will end forcefully (this time with typical error message).

Fixes #22024
Fixes #22254

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible
- [x] Send backports/forwardports if fix needs to be applied to past/future releases
- [x] New public APIs have `@Nonnull/@Nullable` annotations
- [x] New public APIs have `@since` tags in Javadoc

### CODE REVIEW

**Comment by TomaszGaweda** [comment]
run-nightly-tests

**Comment by TomaszGaweda** [comment]
Nightly tests failed due to known issue: https://github.com/hazelcast/hazelcast/issues/19326

**Comment by TomaszGaweda** [comment]
run-nightly-tests

**Comment by TomaszGaweda** [comment]
run-lab-run

**Comment by TomaszGaweda** [comment]
run-lab-run

**Comment by TomaszGaweda** [comment]
run-lab-run

**Review by frant-hartm — CHANGES_REQUESTED** [review]

File: hazelcast/pom.xml [hunk 1]
```diff
@@ -540,12 +540,6 @@
             <optional>true</optional>
         </dependency>
 
-        <dependency>
-            <groupId>com.github.stefanbirkner</groupId>
-            <artifactId>system-lambda</artifactId>
-            <version>1.2.1</version>
-            <scope>test</scope>
-        </dependency>
```
[the remark this answers is not in the document]
Comment (TomaszGaweda) [reply hazelcast/pom.xml:548]: It was added in a branch with previous solution, GitHub seems to be not recognizing this correctly.

in short: I wanted to catch everything from console, this lib seemed to be handy. However, there were some problems that not everything was catched correctly - forks, etc. Our current approach with catching logs is a much improved version with catching via this lib

File: hazelcast/src/test/java/com/hazelcast/jet/SimpleTestInClusterSupport.java [hunk 2]
```diff
@@ -82,6 +83,16 @@ protected static void initializeWithClient(
         client = factory.newHazelcastClient(clientConfig);
     }
 
+    protected void assertNoJobsLeft(HazelcastInstance instance) {
```
[the remark this answers is not in the document]
Comment (TomaszGaweda) [reply hazelcast/src/test/java/com/hazelcast/jet/SimpleTestInClusterSupport.java]: Done 4014327126e453ddc717a22c178ac79acfd36219

File: hazelcast/pom.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (TomaszGaweda) [reply hazelcast/pom.xml:548]: Actually, it was in the same branch: https://github.com/hazelcast/hazelcast/pull/22186/commits/57a2903359d9b4f38cd95a3b778026f2ba546027

[the remark this answers is not in the document]
Comment (frant-hartm) [reply hazelcast/pom.xml:548]: This is in `hazelcast` module, it is also present on master.
Looks like we don't need it anymore, but would be better to remove it in a separate PR.

[the remark this answers is not in the document]
Comment (TomaszGaweda) [reply hazelcast/pom.xml]: Ah, right. Reverted: 2565a403f3a7cdb7816ee5bf1a7188c75b951315

**Comment by TomaszGaweda** [comment]
run-lab-run

**Review by frant-hartm — APPROVED** [review]

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/QueryResultProducerImpl.java [hunk 3]
```diff
@@ -59,8 +55,9 @@ public class QueryResultProducerImpl implements QueryResultProducer {
      * batch jobs where low streaming latency isn't required, but rather we
      * want to return full pages of results to the client.
      */
-    public QueryResultProducerImpl(boolean blockForNextItem) {
-        this.blockForNextItem = blockForNextItem;
+    public QueryResultProducerImpl(boolean isBatch) {
```
Comment (viliam-durina) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/QueryResultProducerImpl.java:58]: What about jobs like this:
```
INSERT INTO a
SELECT * FROM b LIMIT 100;
```
Such job would not use `QueryResultProducer`. I'm not sure we support it now, but we may support it some day.

And the javadoc is stale.

**Review by viliam-durina — CHANGES_REQUESTED** [review]
I don't like the overall design of this PR. Why don't we keep everything as it was, and just reduce the log level to minimum for the `SqlLimitReachedException`? The new impl has multiple flaws.

**Comment by TomaszGaweda** [comment]
Core of the change was merged, the rest needs rework - maybe some day ;) Closing for now

**Pull request closed without merging**